src/Storages/MergeTree/MergeTreeIndexHypothesisMergedCondition.cpp

```diff
@@ -16,199 +21,371 @@ namespace ErrorCodes
             {
-                CNFQuery::AtomicFormula atom{atomic_formula.negative, atomic_formula.ast->clone()};
-                pushNotIn(atom);
+                hypotheses_data.push_back(group);
+                auto atomic_formula = *group.begin();
+                auto atom = cloneAtom(atomic_formula, context);
+
+                assert(!atom.negative);
 
-                const auto * func = atom.ast->as<ASTFunction>();
-                if (func && func->arguments->children.size() == 2)
+                const auto * function_name = functionName(atom);
+                if (function_name && relations.contains(*function_name))
                 {
-                    const auto left = weak_graph->getComponentId(func->arguments->children[0]);
-                    const auto right = weak_graph->getComponentId(func->arguments->children[1]);
-                    if (left && right && weak_graph->hasPath(left.value(), right.value()))
-                    {
-                        useless = false;
-                        return;
-                    }
+                    if constexpr (is_ast)
+                        compare_hypotheses_data.push_back(atom.ast);
+                    else
+                        compare_hypotheses_data.push_back(atom.node_with_hash.node);
                 }
             }
-        });
-    return useless;
-}
+        }
 
-bool MergeTreeIndexhypothesisMergedCondition::mayBeTrueOnGranule(const MergeTreeIndexGranules & granules) const
-{
-    std::vector<bool> values;
-    for (const auto & index_granule : granules)
-    {
-        const auto granule = std::dynamic_pointer_cast<const MergeTreeIndexGranuleHypothesis>(index_granule);
-        if (!granule)
-            throw Exception(ErrorCodes::LOGICAL_ERROR, "Only hypothesis index is supported here.");
-        values.push_back(granule->met);
+        index_to_compare_atomic_hypotheses.push_back(compare_hypotheses_data);
+        index_to_atomic_hypotheses.push_back(hypotheses_data);
     }
 
-    const ComparisonGraph<ASTPtr> * graph = nullptr;
-
+    bool alwaysUnknownOrTrue() const override
     {
-        std::lock_guard lock(cache_mutex);
-        if (const auto it = answer_cache.find(values); it != std::end(answer_cache))
-            return it->second;
+        Nodes active_atomic_formulas(atomic_constraints);
+        for (const auto & hypothesis : index_to_compare_atomic_hypotheses)
+        {
+            active_atomic_formulas.insert(
+                std::end(active_atomic_formulas),
+                std::begin(hypothesis),
+                std::end(hypothesis));
+        }
+
+        auto less_or_equals_resolver = FunctionFactory::instance().get("lessOrEquals", context);
+        auto greater_or_equals_resolver = FunctionFactory::instance().get("greaterOrEquals", context);
+        /// transform active formulas
+        for (auto & formula : active_atomic_formulas)
+        {
+            formula = formula->clone(); /// do all operations with copy
```

Comment:
For Query Tree it'll clone the whole subtree even when we don't need to modify the node. I think it's better to avoid it.

Reply:
I agree, QueryTree is constructed inside this class + when we create `CNF` we clone the nodes again.
And throughout the class I think it's cloned almost always.